app/assets/javascripts/components/projects/ProjectBoard.js

```diff
@@ -1,153 +1,271 @@
 import Stories from "../stories/Stories";
 import Sprints from "../stories/Sprints";
 
 export const ProjectBoard = ({
-  fetchProjectBoard,
   projectId,
-  projectBoard,
-  createStory, 
+  createStory,
   closeHistory,
-  notifications, 
-  removeNotification, 
+  notifications,
+  removeNotification,
   history,
+  projectBoard,
+  fetchProjectBoard,
+  dragDropStory,
   chillyBinStories,
   backlogSprints,
   doneSprints,
   fetchPastStories,
   toggleColumn,
   reverseColumns
 }) => {
+  const [newChillyBinStories, setNewChillyBinStories] = useState();
+  const [newBacklogSprints, setNewBacklogSprints] = useState();
+
+  useEffect(() => {
+    setNewBacklogSprints(backlogSprints);
+  }, [backlogSprints]);
+
   useEffect(() => {
-    fetchProjectBoard(projectId)
+    setNewChillyBinStories(chillyBinStories);
+  }, [chillyBinStories]);
+
+  useEffect(() => {
+    fetchProjectBoard(projectId);
   }, [fetchProjectBoard, projectId]);
 
   if (!projectBoard.isFetched) {
     return <ProjectLoading data-id="project-loading" />;
   }
 
+  const onDragEnd = ({ source, destination, draggableId }) => {
+    const { sprintIndex: sprintDropIndex, columnId: dropColumn } = JSON.parse(
+      destination.droppableId
+    );
+    const { sprintIndex: sprintDragIndex, columnId: dragColumn } = JSON.parse(
+      source.droppableId
+    );
+    const { index: sourceIndex } = source;
+    const { index: destinationIndex } = destination;
+    const isSameColumn = dragColumn === dropColumn;
+    const destinationArray = getArray(
+      dropColumn,
+      newBacklogSprints,
+      newChillyBinStories,
+      sprintDropIndex
+    ); // stories of destination column
+    const sourceArray = getArray(
+      dragColumn,
+      newBacklogSprints,
+      newChillyBinStories,
+      sprintDragIndex
+    ); // stories of source column
+    const dragStory = sourceArray[sourceIndex];
+
+    if (!destination) {
+      return;
+    }
+
+    if (isSameColumn && sourceIndex === destinationIndex) {
+      return;
+    }
+
+    const newPosition = getNewPosition(
+      destinationIndex,
+      sourceIndex,
+      destinationArray,
+      isSameColumn,
+      dragStory.storyType
+    );
+
+    const newStories = moveTask(
+      sourceArray,
+      destinationArray,
+      sourceIndex,
+      destinationIndex
+    );
+
+    // Changing the column array order
+    if (dropColumn === "chillyBin") {
+      setNewChillyBinStories(newStories);
+    }
+
+    if (dropColumn === "backlog") {
+      setNewBacklogSprints(
+        getNewSprints(newStories, newBacklogSprints, sprintDropIndex)
+      );
+    }
+    // Persisting the new array order
+    const newState = getNewState(isSameColumn, dropColumn, dragStory.state);
+    return dragDropStory(dragStory.id, dragStory.projectId, {
+      position: newPosition,
+      state: newState
+    });
+  };
+
+  if (!projectBoard.isFetched) {
+    return <b>{I18n.t("loading")}</b>;
+  }
+
   const columns = [
```

Comment:
As this is array is getting bigger, what do you think about removing it to a new component?

Reply:
Sounds good, the downside is that we have now lots of props to pass down.